corehq/apps/locations/static/locations/js/location_tree.async.js

```diff
@@ -1,410 +1,461 @@
-    this.uuid = ko.observable();
-    this.is_archived = ko.observable();
-    this.can_edit = ko.observable();
-    this.children = ko.observableArray();
-    this.depth = depth || 0;
-    this.children_status = ko.observable('not_loaded');
-    this.expanded = ko.observable(false);
-
-    this.expanded.subscribe(function(val) {
-        if (val === true && (this.children_status() === 'not_loaded' || this.children_status() === 'semi_loaded')) {
-            this.load_children_async();
-        }
-    }, this);
-
-    this.toggle = function() {
-        if (this.expanded() === 'semi') {
-            this.expanded(this.can_have_children());
-        } else {
-            this.expanded(!this.expanded() && this.can_have_children());
-        }
-    };
 
-    this.load = function(data) {
-        this.name(data.name);
-        this.type(data.location_type);
-        this.uuid(data.uuid);
-        this.is_archived(data.is_archived);
-        this.can_edit(data.can_edit);
-        this.expanded(data.expanded);
-        if (data.children_status !== null && data.children_status !== undefined) {
-            this.children_status(data.children_status);
-        }
-        if (data.children != null) {
-            this.set_children(data.children);
-        }
-    };
+        // search for a location within the tree by uuid; return path to location if found
+        this.find_loc = function(uuid, loc) {
+            loc = loc || this.root();
+
+            if (loc.uuid() === uuid) {
+                return [loc];
+            } else {
+                var path = null;
+                $.each(loc.children(), function(i, e) {
+                    var subpath = model.find_loc(uuid, e);
+                    if (subpath) {
+                        path = subpath;
+                        path.splice(0, 0, loc);
+                        return false;
+                    }
+                });
+                return path;
+            }
+        };
+
+        // load location hierarchy
+        this.load = function (locs) {
+            this.root(new LocationModel({
+                name: '_root',
+                children: locs,
+                can_edit: options.can_edit_root,
+                expanded: true,
+                reloadLocationSearchSelect: options.reloadLocationSearchSelect,
+                clearLocationSelection: options.clearLocationSelection,
+                load_locs_url: options.load_locs_url,
+            }, this));
+        };
+    }
+
+    function LocationSearchViewModel(tree_model, options) {
+        // options should have properties:
+        //      "can_edit_root"
+        //      "load_locs_url"
+        //      "reloadLocationSearchSelect"
+        //      "clearLocationSelection"
+
+        var model = this;
+        this.selected_location = ko.observable();
+        this.l__selected_location_id = ko.observable();
+        this.clearLocationSelection = ko.observable(options.clearLocationSelection);
```

Comment:
This probably doesn't need to be an observable. FWIW, it's unlikely that externalizing a page's javascript will involve creating any new observables - and you can always start off by making properties plain js and then switch them to observables if you find you need the extra functionality observables provide.

Reply:
That's a good tip